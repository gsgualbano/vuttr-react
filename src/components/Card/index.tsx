import React, { useState } from 'react';

import Modal from '../Modal';
import RemoveForm from '~/components/Forms/Remove';

import { Tool } from '~/store/ducks/tools/types';

import {
  Container,
  Header,
  Title,
  Description,
  Tags,
  CloseIcon,
  IconContainer,
  Wrapper,
} from './styles';
import DangerButton from '../Common/DangerButton';

/**
 * Props do Card.
 */
interface OwnProps {
  /** objeto com informações sobre a ferramenta como id, title, description, link, tags */
  tool: Tool;
}

/**
 * Card Item
 */
const Card = ({ tool }: OwnProps): JSX.Element => {
  const [showModal, setShowModal] = useState<boolean>(false);

  const handleRemoveButton = (): void => {
    setShowModal(true);
  };

  return (
    <Container data-testid="card-item">
      <Header>
        <Title href={tool.link}>{tool.title}</Title>
        <Wrapper>
          <DangerButton outline onClick={handleRemoveButton}>
            <IconContainer>
              <CloseIcon />
            </IconContainer>
            remove
          </DangerButton>
        </Wrapper>
      </Header>
      <Description>{tool.description}</Description>
      <Tags>
        {tool.tags.map(tag => (
          <p key={tag}>{`#${tag}`}</p>
        ))}
      </Tags>
      <Modal visible={showModal} setShowModal={setShowModal}>
        <RemoveForm setShowModal={setShowModal} tool={tool} />
      </Modal>
    </Container>
  );
};

export default Card;
