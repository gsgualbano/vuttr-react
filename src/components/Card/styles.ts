import styled from 'styled-components';
import { ReactComponent as Icon } from '~/assets/icons/Icon-Close-2px.svg';
import { Theme } from '~/styles/global';

export const Container = styled.div`
  color: ${({ theme }: Theme) => theme.colors.textColor};
  border: 1px solid;
  border-color: ${({ theme }: Theme) => theme.colors.cardBorderColor};
  padding: 10px;
  margin: 20px 0;
  box-shadow: 0 15px 20px 0 rgba(0, 0, 0, 0.07);
  border-radius: 5px;
  background: ${({ theme }: Theme) => theme.colors.cardBackground};
`;

export const Header = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

export const Title = styled.a``;

export const Wrapper = styled.div`
  max-width: 100px;
`;

export const IconContainer = styled.div`
  width: 100%;
  justify-content: center;
  align-items: center;
`;

export const CloseIcon = styled(Icon)`
  width: 8px;
  height: 8px;
  margin-right: 5px;

  g > path.a {
    stroke: #f95e5a;
    stroke-width: 10px;
  }
`;

export const Description = styled.div`
  padding: 15px 0;
`;

export const Tags = styled.div`
  p {
    display: inline-block;
    font-weight: 600;
    margin-right: 4px;
  }
`;
