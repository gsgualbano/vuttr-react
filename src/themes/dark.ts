export default {
  colors: {
    background: '#1B1A1A',
    textColor: '#fff',
    cardBackground: '#29292B',
    cardBorderColor: '#29292B',
    modalBackground: '#29292B',
    modalBorderColor: '#29292B',
    linkColor: '#ae8fd6',
    linkVisitedColor: '##8050bf',
    buttonOutlineBackground: '#29292B',
    InputBackground: '#9e9fa2',
    InputBorderColor: '#9e9fa2',
  },
};
