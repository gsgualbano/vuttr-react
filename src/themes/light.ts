export default {
  colors: {
    background: '#fcfcfd',
    textColor: '#170c3a',
    cardBackground: '#fcfcfd',
    cardBorderColor: '#ebeaed',
    modalBackground: '#fefefe',
    modalBorderColor: '#888',
    linkColor: 'rgb(0, 0, 238)',
    linkVisitedColor: 'rgb(85, 26, 139);',
    buttonOutlineBackground: '#ffffff',
    InputBackground: '#f5f4f6',
    InputBorderColor: '#ebeaed',
  },
};
