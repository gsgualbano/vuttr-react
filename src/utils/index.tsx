/* eslint-disable max-len */
import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import { render } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';
import { reducer } from '../store/ducks/tools';

interface Theme {
  colors: {
    background: string;
    textColor: string;
    cardBackground: string;
    cardBorderColor: string;
    modalBackground: string;
    modalBorderColor: string;
    linkColor: string;
    linkVisitedColor: string;
    buttonOutlineBackground: string;
    InputBackground: string;
    InputBorderColor: string;
  };
}

/** redux mock to render components that depends on the redux */
export const renderWithRedux = (
  component: JSX.Element,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(<Provider store={store}>{component}</Provider>),
  store,
});

/** render components thats depends on the theme provider */
export const renderWithTheme = (component: JSX.Element, theme: Theme) => render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

/** render components thats depends on the redux and theme provider */
export const renderWithReduxAndTheme = (
  component: JSX.Element,
  theme: Theme,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(
    <ThemeProvider theme={theme}>
      <Provider store={store}>{component}</Provider>
    </ThemeProvider>,
  ),
  store,
});
