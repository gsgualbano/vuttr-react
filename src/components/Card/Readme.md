#### Cards com informações sobre as ferramentas cadastradas.

Card Exemplo:

```js
const tool = {
  id: 1,
  title: "Google",
  description: "Search Website",
  link: "https://google.com",
  tags: ["search", "site", "google"]
};

<Card tool={tool} />;
```

Card Dark Exemplo:

```js
import { ThemeProvider } from "styled-components";
import dark from "~/themes/dark";

const tool = {
  id: 1,
  title: "Google",
  description: "Search Website",
  link: "https://google.com",
  tags: ["search", "site", "google"]
};

<ThemeProvider theme={dark}>
  <>
    <Card tool={tool} />
  </>
</ThemeProvider>;
```
