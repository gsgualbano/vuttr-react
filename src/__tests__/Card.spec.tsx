import React from 'react';
import Card from '~/components/Card';

import { render, cleanup, fireEvent } from '@testing-library/react';
import { renderWithReduxAndTheme, renderWithTheme } from '~/utils';

import 'jest-dom/extend-expect';
import 'jest-styled-components';

import dark from '~/themes/dark';
import light from '~/themes/light';

afterEach(cleanup);

describe('<Card />', () => {
  const tool = {
    id: 1,
    title: 'Google',
    description: 'Search Website',
    link: 'https://google.com',
    tags: ['site', 'search', 'engine'],
  };

  it('render correctly', () => {
    const { asFragment } = renderWithTheme(<Card key={tool.id} tool={tool} />, light);
    expect(asFragment()).toMatchSnapshot();
  });

  it('shows modal', () => {
    const { getByText } = renderWithReduxAndTheme(<Card key={tool.id} tool={tool} />, light);
    fireEvent.click(getByText('remove'));
    expect(getByText('Are you sure you want to remove Google ?')).toBeInTheDocument();
  });

  it('Closes modal on cancel button', () => {
    const { queryByText, getByText } = renderWithReduxAndTheme(
      <Card key={tool.id} tool={tool} />,
      light,
    );
    fireEvent.click(getByText('remove'));
    fireEvent.click(getByText('Cancel'));
    expect(queryByText('Are you sure you want to remove Google ?')).not.toBeInTheDocument();
  });

  it('Closes modal on ouside click', () => {
    const { queryByText, getByText, getByTestId } = renderWithReduxAndTheme(
      <Card key={tool.id} tool={tool} />,
      light,
    );
    fireEvent.click(getByText('remove'));
    fireEvent.click(getByTestId('outside-modal'));
    expect(queryByText('Are you sure you want to remove Google ?')).not.toBeInTheDocument();
  });

  it('render with light theme', () => {
    const { getByTestId } = renderWithTheme(<Card key={tool.id} tool={tool} />, light);
    expect(getByTestId('card-item')).toHaveStyleRule('background', '#fcfcfd');
    expect(getByTestId('card-item')).toHaveStyleRule('color', '#170c3a');
    expect(getByTestId('card-item')).toHaveStyleRule('border-color', '#ebeaed');
  });

  it('render with dark theme', () => {
    const { getByTestId } = renderWithTheme(<Card key={tool.id} tool={tool} />, dark);
    expect(getByTestId('card-item')).toHaveStyleRule('background', '#29292B');
    expect(getByTestId('card-item')).toHaveStyleRule('color', '#fff');
    expect(getByTestId('card-item')).toHaveStyleRule('border-color', '#29292B');
  });
});
