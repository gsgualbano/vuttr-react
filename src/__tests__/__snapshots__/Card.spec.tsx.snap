// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> render correctly 1`] = `
<DocumentFragment>
  .c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  color: #f95e5a;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 16px;
  padding: 5px 10px;
  border: none;
  margin-left: 5px;
  background: #ffffff;
  border-radius: 5px;
}

.c3:hover {
  color: #cc4c4c;
}

.c3:click {
  color: #a53f3f;
}

.c0 {
  color: #170c3a;
  border: 1px solid;
  border-color: #ebeaed;
  padding: 10px;
  margin: 20px 0;
  box-shadow: 0 15px 20px 0 rgba(0,0,0,0.07);
  border-radius: 5px;
  background: #fcfcfd;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  max-width: 100px;
}

.c4 {
  width: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  width: 8px;
  height: 8px;
  margin-right: 5px;
}

.c5 g > path.a {
  stroke: #f95e5a;
  stroke-width: 10px;
}

.c6 {
  padding: 15px 0;
}

.c7 p {
  display: inline-block;
  font-weight: 600;
  margin-right: 4px;
}

@media (max-width:768px) {
  .c3 {
    margin-left: 5px;
    width: 100%;
  }
}

<div
    class="c0"
    data-testid="card-item"
  >
    <div
      class="c1"
    >
      <a
        class=""
        href="https://google.com"
      >
        Google
      </a>
      <div
        class="c2"
      >
        <button
          class="c3"
          data-testid="danger-button"
          type="button"
        >
          <div
            class="c4"
          >
            <svg
              class="c5"
            >
              Icon-Close-2px.svg
            </svg>
          </div>
          remove
        </button>
      </div>
    </div>
    <div
      class="c6"
    >
      Search Website
    </div>
    <div
      class="c7"
    >
      <p>
        #site
      </p>
      <p>
        #search
      </p>
      <p>
        #engine
      </p>
    </div>
  </div>
</DocumentFragment>
`;
